### PR TITLE
Maintenance/code cleanup 2026-05 Phase V (performance and pure python resolver prework)

### PR DESCRIPTION
**Pull request by matteius**
## Summary

Phase V of the 2026-05 modernization track. Two intertwined workstreams:

1. **Performance cuts on the existing pip-driven resolver path** — chasing the
   warm-relock and CLI-startup ceilings identified by the May 2026 benchmark
   suite (`benchmarks/benchmark.py`). Final measured win is modest: **~7-11 %**
   across the CI bench (`lock-warm`, `lock-cold`, `install-warm`,
   `add-package`, `import`) from lazy-import work; the architectural ceiling
   is pip's per-package sequential index revalidation and `Link.from_json`
   loop, which pipenv can't move from outside pip.

2. **Initiative G — pure-Python simple-API resolver groundwork** — a new
   in-tree implementation (PEP 691 JSON + PEP 503 HTML client, parsed-manifest
   cache, parallel fetcher) under `pipenv/resolver/` that imports zero
   `pip._internal.*` symbols. Phases 1 + 2 of a four-phase plan ship here;
   Phase 3 (the full `pure-python` backend replacing pip's `PackageFinder`)
   and Phase 4 (promote to default) are explicitly future work.

Companion design + plan docs in `docs/dev/initiative-g-pure-python-design.md`
and `initiative-g-phase1-2-plan.md`. No user-visible API change, no new
runtime dependencies, no lockfile-format change. The opt-in Phase-2 setting
defaults `false`.

---

### Phase-5 perf cuts (lazy imports + empty category skip)

- **`perf(lock): skip resolver subprocess for empty Pipfile categories`** (`49c7a249`) —
  `do_lock` always iterated `["default", "develop", …]` even when a section
  was empty, paying ~5-6 s of subprocess + Resolver/Session setup to produce
  an empty lockfile section. Skip the call when `packages` is an empty
  mapping. Stale entries still cleared because `lockfile.pop(category)`
  already ran.

- **`perf(startup): defer pip-internal network imports in fileutils + internet`** (`7335a1c6`)
  and **`perf(startup): defer pip-internal InstallCommand + unpack + Downloader imports`** (`74a466b1`) —
  the parent CLI and resolver subprocess were both pulling
  `pip._internal.network.download`, `pip._internal.commands.install.InstallCommand`,
  `pip._internal.operations.prepare` (~78 ms each, cumulative) at module
  load. Moved these to first-use inside the functions that need them.
  Tests follow the import-target rename (the prior fix for the same
  symptom on integration tests is at `c85cd3b6`).

- **`perf(resolver): eliminate redundant find_best_candidate walk in resolve_constraints`** (`cf53eb17`) —
  `Resolver.resolve_constraints` called pip's `PackageFinder.find_best_candidate(name, specifier)` once
  per resolved package solely to read `candidate.link.requires_python`.  The resolved tree already
  carries that link (pip's resolvelib stores the chosen candidate on every `InstallRequirement` it
  returns).  Read the marker directly off `result.link`.  Measured on the 100-package bench:
  in-process 31.4 s → 23.4 s (−25.5 %); subprocess warm ~22.6 s → ~17.9 s (−21 %, **~4.7 s saved**).
  Lockfile hash byte-identical for standard pipenv users.

  **Subtle interaction with `tasks/vendoring/patches/patched/pip_finder_ignore_compatability.patch`**
  (commit `3d16ca04` documents this explicitly): users who flip
  `finder._ignore_compatibility = True` in the resolve pipeline (cross-platform locking workflows,
  the patched-pip flag) will see cross-compat packages now CARRY their advertised `requires-python`
  markers in the lockfile.  Pre-fix, the strict `find_best_candidate` returned `None` for those
  candidates and the marker was silently dropped.  This is arguably a correctness fix but IS a
  behavioural change for any consumer that relied on those markers being absent.  A regression-
  pinning test (`test_resolve_constraints_marker_for_ignore_compatibility_link`) and a docstring
  on `resolve_constraints` document the trade-off.

- **`perf(lock): feed prior Pipfile.lock pins as pip constraints on warm relock`** (`20092de9`)
  was reverted (`a2157da4`) after maintainer review caught a semantic
  regression: the change silently froze wildcard versions across
  `pipenv lock` runs, breaking the historical contract that `pipenv lock`
  picks up newer matching versions. The revert is preserved in history
  for traceability.

Measured CI delta vs the pre-phase-5 baseline (median across the bench
suite, single CI run — not multi-run statistical):

| stat              | before    | after     | delta    |
| ----------------- | --------- | --------- | -------- |
| lock-warm         | 21.295 s  | 19.250 s  | −9.6 %   |
| lock-cold         | 25.389 s  | 22.695 s  | −10.6 %  |
| install-warm      | 19.951 s  | 18.648 s  | −6.5 %   |
| install-cold      | 43.191 s  | 41.796 s  | −3.2 %   |
| add-package       | 30.740 s  | 27.405 s  | −10.8 %  |
| import (full)     | 73.263 s  | 69.889 s  | −4.6 %   |

These are real-but-modest gains. The architectural ceiling is documented
in detail in `docs/dev/initiative-g-pure-python-design.md` §2.2.

### Initiative G phases 1 + 2 — pure-Python resolver groundwork

A new `pipenv/resolver/` surface (zero `pip._internal.*` imports, enforced
by a pre-commit grep gate scoped to `pipenv/resolver/`):

- **`pipenv/resolver/candidate.py`** — `Candidate` dataclass + `Hash`
  namedtuple + `Candidate.from_filename` helper. Frozen, slotted, pure
  data; wheel tags derived once at parse time via `pipenv.vendor.packaging.tags`.
- **`pipenv/resolver/pep691_types.py`** — `SimplePageResponse` and
  `FetchError` typed result envelopes.
- **`pipenv/resolver/pep691.py`** — `_parse_pep691_json` (PEP 691 JSON),
  `_parse_pep503_html` (PEP 503 HTML fallback), and `PEP691Client` class.
  Threads per-request `verify` / `cert` into the underlying session
  (FU3, landed in this PR). Deliberately does **not** send
  `Cache-Control: max-age=0` (deliberate divergence from pip).
- **`pipenv/resolver/manifest_cache.py`** — `ParsedManifestCache` with
  TTL, atomic write, schema versioning, and (per FU1, landed here)
  `peek_etag` for stale-cache `If-None-Match` short-circuits.
- **`pipenv/resolver/fetcher.py`** — `ParallelFetcher` with capped 16-worker
  `ThreadPoolExecutor`. Sends `If-None-Match` for stale cache entries
  (FU1); option-a TTL refresh on `304 Not Modified`.
- **`pipenv/resolver/auth.py`** — netrc / URL-embedded basic-auth /
  `PIP_CLIENT_CERT` helpers. Keyring deferred to Phase 3.

Phase-2 integration (opt-in, off by default):

- New setting **`[pipenv] prefetch_index_manifests`** + env-var override
  `PIPENV_PREFETCH_INDEX_MANIFESTS=1`. When enabled,
  `do_lock` calls `_prefetch_index_manifests_if_enabled` to fan out
  per-source parallel pre-fetches (FU2, per-`verify_ssl` policy)
  through pip's own `PipSession` — so pip's `SafeFileCache` is warmed
  as a side effect (no on-disk format reverse engineering).

User-facing doc for the new setting in `docs/pipfile.md`.

#### Why opt-in / why no measured Phase-2 perf claim

T21 (CI bench measurement for the prefetch path) was **explicitly
deferred** during execution review — no appetite for a multi-run
statistically-guarded CI bench step at this time. The Phase-2 perf
hypothesis (parallel cold-cache pre-fetch helps slow networks) is sound
based on the phase-5 I/O analysis but is **theoretical, not measured**
against the current CI baseline. The design doc §11a (`Phase 2a — sign-off note`)
is explicit about this; the user-doc points readers at the phase-5
branch history rather than quoting a percentage.

### Phase-3 follow-ups landed early

Three items originally scoped as Phase-3 work were resolved during the
plan execution after their respective Wave agents flagged them:

- **FU1** (`91c1e4e9`) — `ParsedManifestCache.peek_etag()` + fetcher
  integration. Closes a dead-code path: T19's `status="not-modified"`
  TTL-refresh branch was unreachable before this commit because the
  fetcher never sent `If-None-Match`. Now stale-but-present cache entries
  short-circuit to a 304 instead of re-downloading.
- **FU2** (`4a0ff8a1`) — per-source `verify_ssl` fan-out in
  `_prefetch_index_manifests_if_enabled`. Original T19 cut routed every
  target through the majority-verify session; mixed-policy projects
  (self-signed private index alongside public PyPI) silently fell
  through to pip's cold fetch. Now one `ParallelFetcher` per unique
  `verify_ssl` value. Single-policy projects (common case) unchanged.
- **FU3** (`0047a2e3`) — `PEP691Client.fetch` threads `verify` / `cert`
  into the per-request `session.request(...)` call. Pre-fix these
  kwargs were stored on `self` but never reached the request — so
  FU2's per-source verify routing was effectively a no-op at the
  request layer until FU3 landed.

### Bug fixes caught during execution

- **`c85cd3b6`** — fixes 4 integration tests in `test_import_requirements.py`
  that monkey-patched `pipenv.utils.dependencies.unpack_url` (no longer
  a module attribute after the phase-5 lazy-import work). Patch target
  moved to the canonical source (`pipenv.utils.unpack.unpack_url`).
- **`c76ecb42`** — T20's integration test surfaced a cache-path
  mismatch: T17's `_clear_parsed_manifest_cache` wiped
  `<PIPENV_CACHE_DIR>/manifests-v1/` while T19's prefetch wrote to
  `<PIPENV_CACHE_DIR>/pipenv-manifests/manifests-v1/`. Aligned both on
  the namespaced path.

---

## Test plan

- [x] Full unit suite green locally: **1263 tests passing** (+341 since
      the design-doc commit `a6832ce3`; +307 from Initiative G alone).
- [x] Targeted resolver-module coverage (T17's CI gate, narrow scope):
      **99.68 %** across the six new modules (603 statements, 2 missed
      — the Windows-only `_netrc` branch in `auth.py`, reaches 100 % on
      Windows CI).
- [x] Phase-1 acceptance (T10): byte-equivalent parity vs pip's
      `Link.from_json` / `Link.from_element` across the T2 fixture
      suite (six, django, cryptography, tablib, yanked-pkg, missing-hash
      × JSON + HTML, including a 3496-file `cryptography` snapshot).
      **Zero semantic divergences**; representation diffs documented in
      `tests/unit/test_pep691_parity_known_diffs.md`.
- [x] Phase-2 integration (T20): 5 scenarios passing, 1 skipped
      (self-signed-cert fixture not available — gap documented in
      docstring for a future Phase-3 follow-up).
- [x] FU1+FU2+FU3 each carry their own dedicated tests; coverage
      maintained at 100 % on `candidate.py`, `fetcher.py`,
      `manifest_cache.py`, `pep691.py`, `pep691_types.py`.
- [ ] CI: Lint, build, unit, resolver-module-coverage gate, benchmark.

### Test fixtures added

`tests/unit/fixtures/pep691/*.json` and `tests/unit/fixtures/pep503/*.html`
contain real-PyPI snapshots captured 2026-05-12 plus two hand-crafted
synthetics (`yanked-pkg`, `missing-hash`) exercising edge cases.
Provenance and re-baseline procedure documented in
`tests/unit/fixtures/README.md`.

---

## Companion docs

- `docs/dev/initiative-g-pure-python-design.md` — Initiative G design
  (motivation, architecture, four-phase plan, dependency strategy, open
  questions for sign-off). Status updated to reflect Phase 1 + 2 shipped.
- `initiative-g-phase1-2-plan.md` — the 22-task dependency-aware swarm
  plan that executed Phases 1 + 2 (21 tasks done, T21 deferred, 3
  Phase-3 follow-ups landed early). Each task's `status:` / `log:` /
  `files edited/created:` fields populated post-execution.
- `docs/pipfile.md` — user-facing entry for `[pipenv] prefetch_index_manifests`
  alongside the existing `cool-down-period` block.

News fragments:

- `news/initiative-g-phase1-pep691-client.feature.rst` — Phase 1 ship.
- `news/initiative-g-phase2-prefetch-bridge.feature.rst` — Phase 2 ship.

---

## Migration notes

**None.** Per-commit details:

- All Initiative G modules are new code under `pipenv/resolver/`. No
  existing imports break; the modules are reachable via
  `from pipenv.resolver import PEP691Client, ParsedManifestCache,
  ParallelFetcher, Candidate, Hash, FetchError, SimplePageResponse,
  CachedManifest` if any caller wants them. Phase 1 ships them
  standalone; Phase 2's `do_lock` hook is gated on the opt-in setting.
- Lazy-import changes (`fileutils.py`, `internet.py`, `dependencies.py`,
  `project.py`, `environment.py`, `utils/resolver.py`) keep all
  public attribute names addressable. Tests that monkey-patched
  module-level symbols that became function-scope imports were updated
  to point at the canonical source — the `unpack_url` fix in
  `c85cd3b6` covers the four cases CI flagged. No other tests in the
  repo follow the same anti-pattern (verified by sweep).

## What's NOT in this PR

- **T21 (CI bench measurement for the prefetch path)** — explicitly
  deferred per maintainer scope call. The Phase-2 perf claim is
  theoretical, not measured. Design-doc §11a documents this.
- **Phase 3 — the full `pure_python` backend** that replaces pip's
  `PackageFinder.find_all_candidates` via a `resolvelib.Provider`
  implementation. Three follow-up items (FU1, FU2, FU3) that were
  scoped as Phase-3 work landed here, but the main Phase-3 deliverable
  (the backend itself) is a separate future PR.
- **Self-signed-cert integration test fixture** — T20's skipped
  scenario; needs work in `tests/pytest-pypi/` that's out of scope here.
- **CI bench workflow changes** — the existing benchmark job is
  unchanged. Adding a `PIPENV_PREFETCH_INDEX_MANIFESTS=1` matrix entry
  would require T21-style statistical guards; deferred.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: pipenv/resolver/core.py [hunk 1]
```diff
@@ -547,9 +584,165 @@ def resolve_for_pipenv(request: ResolverRequest) -> ResolverResponse:
...
+    except Exception:  # noqa: BLE001 — defensive against import issues
+        return None
+    try:
+        project = Project()
+        resolver = project.settings.get("resolver")
+    except Exception:  # noqa: BLE001 — any read failure → default
+        return None
+    if not resolver:
+        return None
+    return str(resolver).strip() or None
```
[the remark this answers is not in the document]
Comment (matteius) [reply pipenv/resolver/core.py:649]: @copilot apply changes based on [this feedback](https://github.com/pypa/pipenv/pull/6668#discussion_r3231283898)

**Comment by matteius** [comment]
Restacked onto the updated Phase IV tip and corrected this PR's base from the obsolete `maintenance/code-cleanup-phase4-resolver-followups-2026-05` branch to `maintenance/code-cleanup-phase4-deferred-2026-05`. Range-diff confirms all 46 Phase V commits remain patch-equivalent.

Local verification: 1,283 unit tests passed, 10 platform skips; full pre-commit suite passed.

**Comment by matteius** [comment]
Follow-up: CodeQL flagged the integration test hostname substring assertions. Commit 75fe0a3b now parses each URL and compares exact hostnames; Ruff and diff checks pass. An explicit merge of current main reports Already up to date.

**Comment by matteius** [comment]
Merged the refreshed Phase IV head into Phase V; an explicit current-main merge then reported Already up to date. Full unit suite: 1,284 passed, 11 skipped.

File: tests/unit/test_manifest_cache.py [hunk 2]
```diff
@@ -0,0 +1,1101 @@
...
+  no-litter assertion applies only to the success path.
+* Deterministic serialisation: hashes serialise as sorted lists; wheel_tags
+  serialise as sorted ``str(tag)`` list; the JSON round-trip uses cached
+  ``wheel_tags`` directly so payloads stay bit-stable across
+  ``packaging.tags`` upgrades.
+"""
+from __future__ import annotations
+
+import json
+import os
```
[the remark this answers is not in the document]
Comment (matteius) [reply tests/unit/test_manifest_cache.py]: Fixed in a244c1c2 by removing the unused os import.

File: tests/unit/test_manifest_cache.py [hunk 3]
```diff
@@ -0,0 +1,1101 @@
...
+        cand = _make_wheel_candidate()
+        cache.put(
+            "https://pypi.org/simple", "numpy", [cand], etag=None, ttl_seconds=60,
+        )
+        # Read back: still fresh.
+        assert cache.get("https://pypi.org/simple", "numpy") is not None
+
+        # Now patch ``datetime.now`` inside the module to return a value
+        # 1 hour in the future, well past the 60-second TTL.
+        import pipenv.resolver.manifest_cache as mc
```
[the remark this answers is not in the document]
Comment (matteius) [reply tests/unit/test_manifest_cache.py]: Fixed in a244c1c2. The inline module import is gone; the test now uses pytest monkeypatching with the fully qualified manifest_cache.datetime target.

File: tests/unit/test_manifest_cache.py [hunk 4]
```diff
@@ -0,0 +1,1101 @@
...
+        self, tmp_path, monkeypatch,
+    ):
+        cache = ParsedManifestCache(tmp_path)
+        original = _make_wheel_candidate(name="numpy", version="1.26.0")
+        cache.put("https://pypi.org/simple", "numpy", [original], etag="orig")
+        before = cache.get("https://pypi.org/simple", "numpy")
+        assert before is not None and before.candidates == (original,)
+
+        # Patch os.replace inside the module to always raise OSError.
+        import pipenv.resolver.manifest_cache as mc
```
[the remark this answers is not in the document]
Comment (matteius) [reply tests/unit/test_manifest_cache.py]: Fixed in a244c1c2. The inline module import is gone; os.replace is patched through its fully qualified manifest_cache target.

File: tests/unit/test_manifest_cache.py [hunk 5]
```diff
@@ -0,0 +1,1101 @@
...
+        self, tmp_path, monkeypatch,
+    ):
+        """Cleanup ``os.unlink`` failure must not mask the original OSError.
+
+        Exercises the ``except OSError: pass`` branch in put()'s cleanup.
+        """
+        cache = ParsedManifestCache(tmp_path)
+        cand = _make_wheel_candidate()
+
+        import pipenv.resolver.manifest_cache as mc
```
[the remark this answers is not in the document]
Comment (matteius) [reply tests/unit/test_manifest_cache.py]: Fixed in a244c1c2. The inline module import is gone; os.replace and os.unlink are patched through their fully qualified manifest_cache targets.

File: tests/unit/test_parallel_fetcher.py [hunk 6]
```diff
@@ -0,0 +1,1260 @@
...
+        cache = ParsedManifestCache(tmp_path)
+
+        def fake_fetch(index_url, package_name, *, if_none_match):
+            # `cache.get` will have already barfed with TypeError /
+            # AttributeError on canonicalize_name(None) — but if we got
+            # this far, blow up.
+            raise TypeError("package_name must be str")
+
+        client = _make_client(side_effect=fake_fetch)
+        f = ParallelFetcher(client, cache)
```
[the remark this answers is not in the document]
Comment (matteius) [reply tests/unit/test_parallel_fetcher.py]: Fixed in 68092e8c by removing the unused ParallelFetcher and cache setup from this test.

File: pipenv/resolver/pep691.py [hunk 7]
```diff
@@ -0,0 +1,1054 @@
...
+#: Aggressive enough to fail fast on a flaky mirror; lenient enough for a
+#: slow CI cold-cache run.  T9's retry policy compounds these per-target.
+_DEFAULT_CONNECT_TIMEOUT = 10.0
+_DEFAULT_READ_TIMEOUT = 30.0
+
+#: Accept header advertised on every simple-API request.  Order +
+#: quality values mirror pip's: prefer PEP 691 JSON, accept PEP 691 HTML
+#: at q=0.1 (some private indexes emit JSON only via content negotiation),
+#: accept generic ``text/html`` at q=0.01 (legacy PEP 503-only mirrors).
+_ACCEPT_HEADER = (
```
[the remark this answers is not in the document]
Comment (matteius) [reply pipenv/resolver/pep691.py]: Fixed in 68092e8c by defining the Accept header directly as a PEP691Client class constant.

File: tests/unit/test_pep691_client.py [hunk 8]
```diff
@@ -0,0 +1,1214 @@
...
+class TestAuthHelperEdgeCases:
+    def test_unparseable_host_yields_no_auth_header(self, monkeypatch):
+        """``urllib3.util.parse_url`` raising LocationParseError → no auth.
+
+        Constructed by stubbing ``parse_url`` itself: an invalid URL would
+        normally be rejected earlier in ``extract_url_credentials``, so we
+        force the failure mode the defensive try/except guards against.
+        """
+
+        import pipenv.resolver.pep691 as mod
```
[the remark this answers is not in the document]
Comment (matteius) [reply tests/unit/test_pep691_client.py]: Fixed in 68092e8c by replacing the duplicate module import with fully qualified monkeypatch targets.

File: tests/unit/test_pep691_client.py [hunk 9]
```diff
@@ -0,0 +1,1214 @@
...
+
+        client.fetch("https://host.example/simple", "six")
+
+        headers = _outgoing_headers(session)
+        assert "Authorization" not in headers
+
+    def test_parsed_url_with_empty_host_yields_no_auth(self, monkeypatch):
+        """``parse_url().host is None/""`` short-circuits before netrc lookup."""
+
+        import pipenv.resolver.pep691 as mod
```
[the remark this answers is not in the document]
Comment (matteius) [reply tests/unit/test_pep691_client.py]: Fixed in 68092e8c by replacing the duplicate module import with fully qualified monkeypatch targets.

File: pipenv/resolver/fetcher.py [hunk 10]
```diff
@@ -0,0 +1,459 @@
...
+                ): (idx_url, pkg)
+                for (idx_url, pkg, if_none_match) in to_fetch
+            }
+            for future in concurrent.futures.as_completed(futures):
+                idx_url, package_name = futures[future]
+                # `_fetch_one` catches everything, but defend against a
+                # FutureCancelled or executor-internal failure anyway.
+                try:
+                    fetch_result = future.result()
+                except BaseException as exc:  # noqa: BLE001
```
[the remark this answers is not in the document]
Comment (matteius) [reply pipenv/resolver/fetcher.py]: Fixed in 68092e8c. Executor failures now catch Exception, preserving transient error handling without swallowing process-control exceptions.

File: pipenv/resolver/fetcher.py [hunk 11]
```diff
@@ -0,0 +1,459 @@
...
+        :meth:`PEP691Client.fetch` is already documented as
+        non-raising, but a programming bug (or a test mock that
+        raises) shouldn't crash the worker — so we wrap defensively
+        and map any escapee to a transient :class:`FetchError`.
+        """
+        try:
+            return self._client.fetch(
+                index_url, package_name, if_none_match=if_none_match
+            )
+        except BaseException as exc:  # noqa: BLE001
```
[the remark this answers is not in the document]
Comment (matteius) [reply pipenv/resolver/fetcher.py]: Fixed in 68092e8c. Worker failures now catch Exception, and regression coverage verifies KeyboardInterrupt propagates.

File: tests/unit/test_manifest_cache.py [hunk 12]
```diff
@@ -0,0 +1,1100 @@
...
+                        if stop_event.is_set():
+                            return
+                        cache.put(
+                            "https://pypi.org/simple",
+                            "numpy",
+                            payload,
+                            etag=f"v-{payload[0].version}",
+                        )
+                        time.sleep(0.001)
+            except BaseException as exc:  # pragma: no cover
```
[the remark this answers is not in the document]
Comment (matteius) [reply tests/unit/test_manifest_cache.py]: Fixed in 68092e8c by narrowing the test-thread error collector to Exception.

File: tests/unit/test_manifest_cache.py [hunk 13]
```diff
@@ -0,0 +1,1100 @@
...
+                    # interleaved oddly) or a fully-valid CachedManifest.
+                    if got is not None:
+                        assert isinstance(got, CachedManifest)
+                        # Touch each field to ensure full parse worked.
+                        assert isinstance(got.candidates, tuple)
+                        assert all(
+                            isinstance(c, Candidate) for c in got.candidates
+                        )
+                        reads_succeeded[0] += 1
+            except BaseException as exc:
```
[the remark this answers is not in the document]
Comment (matteius) [reply tests/unit/test_manifest_cache.py]: Fixed in 68092e8c by narrowing the test-thread error collector to Exception.

File: tests/unit/test_manifest_cache.py [hunk 14]
```diff
@@ -0,0 +1,1100 @@
...
+        barrier = threading.Barrier(2)
+        errors: list[BaseException] = []
+
+        def writer(cand: Candidate, etag: str):
+            try:
+                barrier.wait(timeout=5)
+                cache.put(
+                    "https://pypi.org/simple", "numpy", [cand], etag=etag,
+                )
+            except BaseException as exc:  # pragma: no cover
```
[the remark this answers is not in the document]
Comment (matteius) [reply tests/unit/test_manifest_cache.py]: Fixed in 68092e8c by narrowing the test-thread error collector to Exception.

File: pipenv/routines/lock.py [hunk 15]
```diff
@@ -9,6 +10,257 @@
...
+
+    def _safe_populate(fetcher, group_targets):
+        # Best-effort per-policy populate: a per-fetcher exception must
+        # not abort the OTHER fetchers' populate.  Extracted from the
+        # loop body so the ``try``/``except`` isn't flagged by ruff
+        # PERF203 — same swallow-and-continue semantics as T19, just
+        # applied per-policy.
+        try:
+            fetcher.populate(group_targets)
+        except Exception:
```
[the remark this answers is not in the document]
Comment (matteius) [reply pipenv/routines/lock.py:241]: Fixed in 68092e8c with an explanatory comment documenting why optional manifest prefetch failures are intentionally ignored.

File: pipenv/resolver/manifest_cache.py [hunk 16]
```diff
@@ -0,0 +1,505 @@
...
+            finally:
+                tmp.close()
+            _replace_with_windows_retry(tmp.name, str(target))
+        except BaseException:
+            # os.replace failed (or any earlier step did).  Best-effort
+            # cleanup of the temp file — if this also fails (rare), the
+            # leftover is named ``<target>.<random>.tmp`` and harmless.
+            try:
+                os.unlink(tmp.name)
+            except OSError:
```
[the remark this answers is not in the document]
Comment (matteius) [reply pipenv/resolver/manifest_cache.py:299]: Fixed in d1248c21. The handler now documents that cleanup failure must not replace the original write failure and that the uniquely named leftover temp file is harmless.

File: pipenv/resolver/manifest_cache.py [hunk 17]
```diff
@@ -0,0 +1,505 @@
...
+
+    def invalidate(self, index_url: str, package_name: str) -> None:
+        """Best-effort removal of a single cache entry.
+
+        Missing-file is not an error.
+        """
+        target = self._path_for(index_url, package_name)
+        try:
+            target.unlink()
+        except FileNotFoundError:
```
[the remark this answers is not in the document]
Comment (matteius) [reply pipenv/resolver/manifest_cache.py:313]: Fixed in d1248c21. The FileNotFoundError handler now explicitly documents that an already-absent cache entry is the intended end state.

**Comment by matteius** [comment]
Rebased the full Phase V series onto current `main` after #6667 and #6611 merged, and addressed both new manifest-cache code-quality findings in `d1248c21`.

Range-diff confirms all 51 commits remain patch-equivalent apart from that review fix. Local validation: `1,300 passed, 11 skipped`; full pre-commit suite passed.

**Comment by matteius** [comment]
Addressed the three suppressed review findings in `5cdd812a`:

- prefetch now runs only when `prefetch_index_manifests` resolves to the literal boolean `True`; an invalid value such as `PIPENV_PREFETCH_INDEX_MANIFESTS=maybe` safely leaves it disabled
- removed the unused duplicate `Environment.PIPENV_PREFETCH_INDEX_MANIFESTS` attribute; `Settings.get()` remains the single env/Pipfile precedence path
- replaced the POSIX-only `bash`/`grep` import guard with pre-commit's portable `pygrep` language
- added regression coverage for the invalid environment value and the prefetch short-circuit

Validation:
- `python -m pytest -o addopts='' -q tests/unit --ignore=tests/unit/test_release_tasks.py --ignore=tests/unit/test_vendor.py` → 1302 passed, 11 skipped
- `pre-commit run --all-files --verbose --show-diff-on-failure` → passed
